client/ipa-join.c

```diff
@@ -612,6 +616,231 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    struct curl_buffer cb;
+
+    json_t *json_post = NULL;
+    char *json_post_str = NULL;
+
+    json_t *j_root = NULL;
+    json_t *j_result = NULL;
+    json_t *j_props = NULL;
+
+    json_error_t j_error;
+
+    int rval = 0;
+
+    *hostdn = NULL;
+    *princ = NULL;
+
+    uname(&uinfo);
+
+    if (hostname == NULL) {
+        host = strdup(uinfo.nodename);
+    } else {
+        host = strdup(hostname);
+    }
+
+    curl_global_init(CURL_GLOBAL_DEFAULT);
+
+    curl = curl_easy_init();
+    if (!curl) {
+            curl_global_cleanup();
+            return 1;
+    }
+
+    cb.payload = (char*) malloc(sizeof(cb.payload));
+    if (cb.payload == NULL) {
+        if (!quiet)
+            fprintf(stderr, _("Out of memory!\n"));
+        rval = 3;
+        goto cleanup;
+    }
+
+    cb.size = 0;
+
+#define CURL_SETOPT(curl, opt, val) \
+    if (curl_easy_setopt(curl, opt, val) != CURLE_OK) { \
+        rval = 17; \
+        goto cleanup; \
+    }
+
+#define ASPRINTF(strp, fmt...) \
+    if (asprintf(strp, fmt) == -1) { \
+        if (!quiet) \
+            fprintf(stderr, _("Out of memory!\n")); \
+        rval = 3; \
+        goto cleanup; \
+    }
+
+    /* setting endpoint and custom headers */
+    ASPRINTF(&url, "https://%s/ipa/json", ipaserver);
+    CURL_SETOPT(curl, CURLOPT_URL, url);
+
+    ASPRINTF(&referer, "referer: https://%s/ipa", ipaserver);
+    headers = curl_slist_append(headers, referer);
+    ASPRINTF(&user_agent, "User-Agent: %s/%s", NAME, VERSION);
+    headers = curl_slist_append(headers, user_agent);
+
+    headers = curl_slist_append(headers, "Accept: application/json");
+    headers = curl_slist_append(headers, "Content-Type: application/json");
+    CURL_SETOPT(curl, CURLOPT_HTTPHEADER, headers);
+
+    CURL_SETOPT(curl, CURLOPT_CAINFO, DEFAULT_CA_CERT_FILE);
+
+    CURL_SETOPT(curl, CURLOPT_WRITEFUNCTION, &jsonrpc_handle_response);
+    CURL_SETOPT(curl, CURLOPT_WRITEDATA, &cb);
+
+    /* delegate authentication to GSSAPI */
+    CURL_SETOPT(curl, CURLOPT_GSSAPI_DELEGATION, CURLGSSAPI_DELEGATION_FLAG);
+    CURL_SETOPT(curl, CURLOPT_HTTPAUTH, CURLAUTH_NEGOTIATE);
+    CURL_SETOPT(curl, CURLOPT_USERPWD, ":");
+
+    if (debug)
+        CURL_SETOPT(curl, CURLOPT_VERBOSE, 1L);
+
+    /* create the JSON-RPC payload */
+    json_post = json_pack_ex(&j_error, 0, "{s:s, s:[[s], {s:s, s:s}]}", "method", "join", "params", host, "nsosversion",
+                             uinfo.release, "nshardwareplatform", uinfo.machine);
+
+    if (json_post == NULL) {
+        if (debug != 0)
+            fprintf(stderr, _("json_pack failed: %s\n"), j_error.text);
+
+        rval = 17;
+        goto cleanup;
+    }
+
+    json_post_str = json_dumps(json_post, 0);
+    CURL_SETOPT(curl, CURLOPT_POSTFIELDS, json_post_str);
+
+    if (debug != 0)
+        fprintf(stderr, _("JSON-RPC request:\n%s\n"), json_post_str);
+
+    /* Perform the call and check for errors */
+    res = curl_easy_perform(curl);
+    if (res != CURLE_OK)
+    {
```

Comment:
This does not react to expired tickets properly. It instead tries to parse the output of the returned login page as json and fails. The output of ipa-join -d contains:

```
> POST /ipa/json HTTP/1.1
Host: ipa.example.test
referer: https://ipa.example.test/ipa
User-Agent: ipa-join/4.9.0.dev202006011924+gita7b425d68
Accept: application/json
Content-Type: application/json
Content-Length: 131

* upload completely sent off: 131 out of 131 bytes
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 401 Unauthorized
< Date: Mon, 01 Jun 2020 19:46:03 GMT
< Server: Apache/2.4.41 (Fedora) OpenSSL/1.1.1d mod_wsgi/4.6.4 Python/3.7 mod_auth_gssapi/1.6.1
* gss_init_sec_context() failed: Ticket expired. 
< WWW-Authenticate: Negotiate
< X-Frame-Options: DENY
< Content-Security-Policy: frame-ancestors 'none'
< Last-Modified: Fri, 01 May 2020 17:21:53 GMT
< Accept-Ranges: bytes
< Content-Length: 2714
< Content-Type: text/html; charset=UTF-8
```
I think that in addition to the if (res != CURLE_OK) check you'll need to check the HTTP response code.

To test do: 
```
# kinit -l 1 admin
# ipa-join -d
```

Reply:
Fixed.